backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala

```diff
@@ -362,30 +365,58 @@ trait StandardAsyncExecutionActor extends AsyncBackendJobExecutionActor with Sta
-    this will need to be refactored.
-     */
 
-    def validateAdHocFile(value: WomValue): ErrorOr[List[WomFile]] = value match {
-      case womFile: WomFile => List(womFile).valid
-      case womArray: WomArray => womArray.value.toList.traverse(validateAdHocFile).map(_.flatten)
-      case other =>
-        s"Ad-hoc file creation expression invalidly created a ${other.womType.toDisplayString} result.".invalidNel
+  /**
+    * By default, ad hoc values get localized to the call directory.
+    * This way if running locally with docker they get mounted with the rest of the inputs in the container.
+    * The PAPI backend overrides this to a noop since the localization happens on the VM directly, so there's no need
+    * for this extra localization step.
+    * 
+    * Maybe this should be the other way around: the default implementation is noop and SFS / TES override it ?
+    */
+  lazy val localizeAdHocValues: List[AdHocValue] => ErrorOr[List[StandardAdHocValue]] = { adHocValues => 
+    import cats.instances.future._
+
+    // Localize an adhoc file to the callExecutionRoot as needed
+    val localize: (AdHocValue, Path) => Future[StandardAdHocValue] = {
+      // If the original file is not already under callExecutionRoot, then copy it there
+      case (adHoc @ AdHocValue(_, None, _), file) if !file.isChildOf(jobPaths.callExecutionRoot) =>
+        asyncIo.copyAsync(file, jobPaths.callExecutionRoot / file.name) as {
+          Coproduct[StandardAdHocValue](LocalizedAdHocValue(adHoc, jobPaths.callExecutionRoot / file.name))
+        }
+      // If the file should be available as a different name than its own, copy it as that
+      case (adHoc @ AdHocValue(_, Some(alternateName), _), file) =>
+        val finalPath = jobPaths.callExecutionRoot / alternateName
+        // First check that it's not already there under execution root
+        asyncIo.existsAsync(finalPath) flatMap {
+          // If it's not then copy it
+          case false => asyncIo.copyAsync(file, finalPath).map(_ => Coproduct[StandardAdHocValue](LocalizedAdHocValue(adHoc, finalPath)))
+          case true => Future.successful(Coproduct[StandardAdHocValue](LocalizedAdHocValue(adHoc, finalPath)))
+        }
+      case (adHoc, path) => Future.successful(Coproduct[StandardAdHocValue](LocalizedAdHocValue(adHoc, path)))
     }
+    
+    adHocValues.traverse[ErrorOr, (AdHocValue, Path)]({ adHocValue =>
+      // Build an actionable Path from the ad hoc file
+      getPath(adHocValue.womValue.value).toErrorOr.map(adHocValue -> _)
+    })
+      // Localize the values if necessary
+      .map(_.traverse[Future, StandardAdHocValue](localize.tupled)).toEither
+      // TODO: Asynchronify
+      // This is obviously sad but turning it into a Future has earth-shattering consequences, so synchronizing it for now
+      .flatMap(future => Try(Await.result(future, 1.hour)).toChecked)
```

Comment:
Wait, no this is the real 🙈

Reply:
you might want to save up your 🙈 for my next PR